Hudl.Mjolnir/Command/BulkheadInvoker.cs

```diff
@@ -0,0 +1,127 @@
+            var bulkhead = _context.GetBulkhead(command.BulkheadKey);
+
+            if (!bulkhead.TryEnter())
+            {
+                _context.MetricEvents.RejectedByBulkhead(bulkhead.Name, command.Name);
+                throw new BulkheadRejectedException();
+            }
+
+            _context.MetricEvents.EnterBulkhead(bulkhead.Name, command.Name);
+
+            // This stopwatch should begin stopped (hence the constructor instead of the usual
+            // Stopwatch.StartNew(). We'll only use it if we aren't using circuit breakers.
+            var stopwatch = new Stopwatch();
+            var executedHere = false;
+            try
+            {
+                if (_useCircuitBreakers.Value)
+                {
+                    executedHere = false;
+                    return await _breakerInvoker.ExecuteWithBreakerAsync(command, ct).ConfigureAwait(false);
+                }
+
+                executedHere = true;
+                stopwatch.Start();
+                return await command.ExecuteAsync(ct).ConfigureAwait(false);
+            }
+            finally
+            {
+                bulkhead.Release();
+
+                _context.MetricEvents.LeaveBulkhead(bulkhead.Name, command.Name);
+
+                if (executedHere)
+                {
+                    stopwatch.Stop();
+                    command.ExecutionTimeMillis = stopwatch.Elapsed.TotalMilliseconds;
+                }
+            }
+        }
+
+        public TResult ExecuteWithBulkhead<TResult>(SyncCommand<TResult> command, CancellationToken ct)
+        {
+            var bulkhead = _context.GetBulkhead(command.BulkheadKey);
+
+            if (!bulkhead.TryEnter())
+            {
+                _context.MetricEvents.RejectedByBulkhead(bulkhead.Name, command.Name);
+                throw new BulkheadRejectedException();
+            }
+
+            _context.MetricEvents.EnterBulkhead(bulkhead.Name, command.Name);
+
+            // This stopwatch should begin stopped (hence the constructor instead of the usual
+            // Stopwatch.StartNew(). We'll only use it if we aren't using circuit breakers.
+            var stopwatch = new Stopwatch();
+            var executedHere = false;
+            try
+            {
+                if (_useCircuitBreakers.Value)
+                {
+                    executedHere = false;
+                    return _breakerInvoker.ExecuteWithBreaker(command, ct);
+                }
+
+                executedHere = true;
+                stopwatch.Start();
+                return command.Execute(ct);
+            }
+            finally
+            {
+                bulkhead.Release();
+
+                _context.MetricEvents.LeaveBulkhead(bulkhead.Name, command.Name);
+
+                if (executedHere)
+                {
+                    stopwatch.Stop();
+                    command.ExecutionTimeMillis = stopwatch.Elapsed.TotalMilliseconds;
+                }
+            }
+        }
+    }
+}
```

Comment:
Any suggestions? I've thought through and tried a handful of different things, and all of them end up not working. I'd _love_ to get rid of the duplication, though. We solved this in the old Command path by doing an async-to-sync conversion in the library, but getting rid of that is one of the goals of this PR.